ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -178,102 +183,208 @@ newtype Our a = Our { unOur :: a }
-    cschKill       :: !(m ())
-    -- | Latest tip announced by the remote peer
-  , cschTheirTip   :: !(STM m (Maybe (Tip blk)))
-    -- | Slot of the last received header
-  , cschLatestSlot :: !(STM m SlotNo)
+    cschKill  :: !(m ())
+
+    -- | Data shared between the client and external components like GDD.
+  , cschState :: !(StrictTVar m (ChainSyncState blk))
+  }
+  deriving stock (Generic)
+
+deriving anyclass instance (
+  IOLike m,
+  HasHeader blk,
+  NoThunks (Header blk)
+  ) => NoThunks (ChainSyncClientHandle m blk)
+
+-- | Convenience function for reading a nested set of TVars and extracting some
+-- data from 'ChainSyncState'.
+viewChainSyncState ::
+  IOLike m =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  (ChainSyncState blk -> a) ->
+  STM m (Map peer a)
+viewChainSyncState varHandles f =
+  Map.map f <$> (traverse (readTVar . cschState) =<< readTVar varHandles)
+
+-- | Convenience function for reading the 'ChainSyncState' for a single peer
+-- from a nested set of TVars.
+chainSyncStateFor ::
+  Ord peer =>
+  IOLike m =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  peer ->
+  STM m (ChainSyncState blk)
+chainSyncStateFor varHandles peer =
+  readTVar . cschState . (Map.! peer) =<< readTVar varHandles
+
+-- | Interface for the ChainSync client to manipulate the idling flag in
+-- 'ChainSyncState'.
+data Idling m = Idling {
+    -- | Mark the peer as being idle.
+    idlingStart :: !(m ())
+
+    -- | Mark the peer as not being idle.
+  , idlingStop  :: !(m ())
+  }
+  deriving stock (Generic)
+
+deriving anyclass instance IOLike m => NoThunks (Idling m)
+
+-- | No-op implementation, for tests.
+noIdling :: Applicative m => Idling m
+noIdling =
+  Idling {
+      idlingStart = pure ()
+    , idlingStop  = pure ()
+    }
+
+-- | Interface to the LoP implementation for the ChainSync client.
+data LoPBucket m = LoPBucket {
+    -- | Pause the bucket, because the peer is alert and we're waiting for some
+    -- condition.
+    lbPause      :: !(m ())
+
+    -- | Resume the bucket after pausing it.
+  , lbResume     :: !(m ())
+
+    -- | Notify the bucket that the peer has sent an interesting header.
+  , lbGrantToken :: !(m ())
   }
   deriving stock (Generic)
-  deriving (NoThunks) via AllowThunk (ChainSyncClientHandle m blk)
 
+deriving anyclass instance IOLike m => NoThunks (LoPBucket m)
+
+-- | No-op implementation, for tests.
+noLoPBucket :: Applicative m => LoPBucket m
+noLoPBucket =
+  LoPBucket {
+      lbPause      = pure ()
+    , lbResume     = pure ()
+    , lbGrantToken = pure ()
+    }
+
+-- | Interface for the ChainSync client to its state allocated by
+-- 'bracketChainSyncClient'.
+data ChainSyncStateView m blk = ChainSyncStateView {
+    -- | The current candidate fragment
+    csvSetCandidate  :: !(AnchoredFragment (Header blk) -> STM m ())
+
+    -- | Update the slot of the latest received header
+  , csvSetLatestSlot :: !(WithOrigin SlotNo -> STM m ())
```

Comment:
Hmm, this interfaces means we will have separate `modifyTVar` calls even if we invoke `csvSetCandidate` and `csvSetLatestSlot` right next to each other, compared to exposing `(ChainSyncState blk -> ChainSyncState blk) -> STM m ()`. However, this doesn't seem major.